crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,448 @@
+        }
+    }
+}
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[cfg(test)]
+    pub(crate) fn with_id(id: u32) -> Self {
+        Self {
+            id,
+            nodes: Default::default(),
+        }
+    }
+
+    /// Creates a root graph node without any dependencies. A graph can have multiple distinct
+    /// root nodes.
+    pub fn root<Params>(&self, system: impl IntoSystemDescriptor<Params>) -> SystemGraphNode {
+        self.create_node(system.into_descriptor())
+    }
+
+    /// Checks if two graphs instances point to the same logical underlying graph.
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let mut nodes = self.nodes.borrow_mut();
+        assert!(
+            nodes.len() <= u32::MAX as usize,
+            "Cannot add more than {} systems to a SystemGraph",
+            u32::MAX
+        );
+        let id = NodeId(self.id, nodes.len() as u32);
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        nodes.insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.nodes.borrow_mut().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion to [SystemSet]. All other clones of the same graph will be empty
+/// afterwards.
+///
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
+    fn from(graph: SystemGraph) -> Self {
+        let mut system_set = SystemSet::new();
+        for (_, system) in graph.nodes.borrow_mut().drain() {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => {
+                    system_set = system_set.with_system(descriptor);
+                }
+                SystemDescriptor::Exclusive(descriptor) => {
+                    system_set = system_set.with_system(descriptor);
+                }
+            }
+        }
+        system_set
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemGraphNode {
+    id: NodeId,
+    graph: SystemGraph,
+}
+
+impl SystemGraphNode {
+    pub fn graph(&self) -> SystemGraph {
+        self.graph.clone()
+    }
+
+    /// Creates a new node in the graph and adds the current node as it's dependency.
```

Comment:
```suggestion
    /// Creates a new node in the graph and adds the current node as its dependency.
```
Yeah, English language is weird. Pronouns don't get a posessive apostrophe (his, hers, its).